lib/publisher/sqsapppublisher.go

```diff
@@ -0,0 +1,66 @@
+package publisher
+
+import (
+	"encoding/json"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sqs"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+type sqsAppPublisherService struct {
+	sqsService *sqs.SQS
+	config     sqsPublisherConfig
+}
+
+type sqsPublisherConfig struct {
+	QueueName string
+}
+
+func NewSQSAppPublisherService(queueName string) AppPublisher {
+	log.Info("Initializing the sqs app publisher service")
+	session := session.Must(session.NewSession(&aws.Config{
+		Region: aws.String("us-west-2"),
+	}))
+
+	sqsService := sqs.New(session)
+	config := sqsPublisherConfig{
+		QueueName: queueName,
+	}
+	log.Info("Initialized the sqs app publisher service")
+	return &sqsAppPublisherService{sqsService, config}
+}
+
+func (s *sqsAppPublisherService) Publish(event AppChangeEvent) error {
+	log.Infof("Publishing event for tenant: %v and app type: %v", event.Tenant, event.AppType)
+	queueUrlOutput, err := s.sqsService.GetQueueUrl(&sqs.GetQueueUrlInput{
+		QueueName: &s.config.QueueName,
+	})
+	if err != nil {
+		log.Errorf("Failed to retrieve the qeueue url:%v err: %v", s.config.QueueName, err)
+		return trace.Wrap(err)
+	}
+	messageJsonData, err := json.Marshal(event)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	log.Infof("Publishing the message for app type: %v", event.AppType)
+	_, err = s.sqsService.SendMessage(&sqs.SendMessageInput{
+		MessageDeduplicationId: aws.String(createMessageDeduplicationId(event)),
+		MessageBody:            aws.String(string(messageJsonData)),
+		QueueUrl:               queueUrlOutput.QueueUrl,
+	})
+
+	if err != nil {
+		log.Errorf("Failed to publish the message for app type: %v  err : %v", event.AppType, err)
+		return err
+	}
+	log.Infof("Sent message for app type: %v successfully", event.AppType)
+	return nil
+}
+
+func createMessageDeduplicationId(event AppChangeEvent) string {
+	return event.Tenant + "##" + string(event.AppType)
```

Comment:
This is how SQS deduplication works.

1. A ssh node was activated and message was sent : `{ "tenant": "dev", "type": "server" }`
2. This message is quickly processed by app management.
3. Another SSH server got deleted within 5 minute (de-dup window) and message was publish: `{ "tenant": "dev", "type": "server" }`. SQS will automatically drop this message because deduplication id for both is same.

Generate random number so that messages are not considered duplicates.

Reply:
Sending UUID for every notification